### PR TITLE
Add CircomProofs for Addition & Update

### PR DESCRIPTION
**Pull request by arnaucube**
Add CircomProcessorProof & CircomVerifierProof for Addition and Updates on the MerkleTree.

### CODE REVIEW

File: merkletree.go [hunk 1]
```diff
@@ -305,62 +346,66 @@ func (mt *MerkleTree) addNode(tx db.Tx, n *Node) (*Hash, error) {
...
-	return nil, ErrKeyNotFound
+	return nil, nil, nil, ErrKeyNotFound
 }
 
 // Update updates the value of a specified key in the MerkleTree, and updates
-// the path from the leaf to the Root with the new values.
-func (mt *MerkleTree) Update(k, v *big.Int) error {
+// the path from the leaf to the Root with the new values. Returns the
+// CircomProcessorProof.
+func (mt *MerkleTree) Update(k, v *big.Int) (*CircomProcessorProof, error) {
```
Comment (arnaubennassar) [inline merkletree.go:393]: I think it should be `UpdateAndGetCircomProof` just like `AddAndGetCircomProof`.
Looking at the entire code I've seen that there is two "adds", one thar return a CricomProofProcessor amd one that doesnt. I think that the Update function should do the same

**Review by arnaubennassar — APPROVED** [review]
Have a look at the suggestion

File: merkletree.go (hunk 1 above)
Comment (arnaucube) [reply merkletree.go:393]: The thing is that the `AddAndGetCircomProof` has extra computation from `Add` pure method, while `Update` does not get much extra computation returning the CircomProof or not.
That's why there are 2 methods for Addition, one without returning the CircomProof for the cases where is not needed, and another one returning the CircomProof with the extra computation that requires. While for Update, it does not need much extra computation to generate the CircomProof, and that's why I've kept in a single method (without having an 'update without CircomProof/ update with CircomProof'), where if not needed the CircomProof can be ignored without having much unused extra computation.
In other words, the `AddAndGetCircomProof` is an extra wrapper to `Add`, adding that extra computation to generate the CircomProof, while `Update` does not need the extra computation to return the CircomProof. In the future (when we may have extra time), `Delete` method will do similar to the `Update` method, returning the CircomProof.
